crates/model/src/order.rs

```diff
@@ -391,6 +344,78 @@ impl OrderCreation {
         });
         signing::keccak256(&hash_data)
     }
+
+    pub fn token_pair(&self) -> Option<TokenPair> {
+        TokenPair::new(self.buy_token, self.sell_token)
+    }
+
+    pub fn uid(&self, domain: &DomainSeparator, owner: &H160) -> OrderUid {
+        let mut uid = OrderUid([0u8; 56]);
+        uid.0[0..32].copy_from_slice(&super::signature::hashed_eip712_message(
+            domain,
+            &self.hash_struct(),
+        ));
+        uid.0[32..52].copy_from_slice(owner.as_fixed_bytes());
+        uid.0[52..56].copy_from_slice(&self.valid_to.to_be_bytes());
+        uid
+    }
+}
+
+// An order as provided to the orderbook by the frontend.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OrderCreation {
+    #[serde(flatten)]
+    pub data: OrderData,
+    #[serde(flatten)]
+    pub signature: CreationSignature,
+    pub quote_id: Option<QuoteId>,
+}
+
+impl OrderCreation {
+    /// Recovers the owner address for the specified domain.
+    ///
+    /// This can return an error for orders with ECDSA signatures if the
+    /// optionally specified `from` address does not match the address
+    /// EC-recovered address from the signature.
+    ///
+    /// This method will also return an error if the Signature is `EIP-1271`,
+    /// since it's validated on-chain.
+    pub fn recover_owner(&self, domain: &DomainSeparator) -> Result<H160, RecoveryError> {
+        let (scheme, from, signature) = match &self.signature {
+            CreationSignature::Eip712 { from, signature } => (EcdsaSigningScheme::Eip712, from, signature),
+            CreationSignature::EthSign { from, signature } => (EcdsaSigningScheme::EthSign, from, signature),
+            CreationSignature::Eip1271 { .. } => return Err(RecoveryError::OnChainRecovery),
```

Comment:
Why can't we return the `from` address here? While we need to verify the signature on-chain, I would argue we can recover the signer address trivially:

```suggestion
            CreationSignature::Eip1271 { from, .. } => return Ok(*from),
```

Reply:
Alternatively, we can make `recover_owner` `async` and then do the `eth_call` here.

Reply:
> Alternatively, we can make `recover_owner` `async` and then do the `eth_call` here.

For `OrderCreation` to have an async call to do `eth_call` it will need to know about the service that will do the call itself, which might complicate this struct, especially since it's just a struct for payload deserialized from `json`.
I guess we can pass the service but then we will be mixing 2 different responsibilities

Reply:
We could pass in a `web3` instance to `recover_signature`, but I agree its a bit icky.

Reply:
Yes, we either need to use a separate trait and implementation for Web3, same as we have in the `orderbook` or we should perform the validation of the owner on creation.
Somehow it feels off to have CreateSignature recover owner & return an error when the recovered address doesn't match the passed `from` address and also do the same kind of validation in another place (for EIP-1271) using the contract.

Calling the contract's `isValidSignature` at the `from` address and failing to return the Magic value to me is the same thing as failing to recover the owner.

What do you think? Any new ideas how we can handle the recovery and validation?
Maybe we can recover the owner (as you mentioned) by returning the `from` address in the `CreateSignature` but then matching against it and validating the signatures on spot in the API instead of a `validate` method?